src/graphblas/bsp1d/init.cpp

```diff
@@ -130,163 +136,271 @@ grb::RC grb::internal::BSP1D_Data::initialize(
 	payload_size = 0;
 	tag_size = 0;
 	max_msgs = 0;
+	cur_payload = nullptr;
+	cur_msgs = nullptr;
+	queue_status = 0;
+	queue = LPF_INVALID_BSMP;
+	coll = LPF_INVALID_COLL;
+	// end default initialisers
+
+	// try allocations (and initialisations) first
 	cur_payload = new( std::nothrow ) size_t[ P ];
-	if( cur_payload == NULL ) {
-#ifdef _DEBUG
-		std::cerr << "\t out of memory while initializing (1)\n";
-#endif
+	if( cur_payload == nullptr ) {
+		std::cerr << "\tError during BSP1D::init: "
+			<< "out of memory while initializing (1)\n";
 		return OUTOFMEM;
 	}
 	cur_msgs = new( std::nothrow ) size_t[ P ];
-	if( cur_msgs == NULL ) {
-		delete[] cur_payload;
-#ifdef _DEBUG
-		std::cerr << "\t out of memory while initializing (2)\n";
-#endif
+	if( cur_msgs == nullptr ) {
+		delete [] cur_payload;
+		std::cerr << "\tError during BSP1D::init: "
+			<< "out of memory while initializing (2)\n";
 		return OUTOFMEM;
 	}
 	for( size_t i = 0; i < P; ++i ) {
 		cur_payload[ i ] = cur_msgs[ i ] = 0;
 	}
-	queue_status = 0;
-	queue = LPF_INVALID_BSMP;
 
 	if( _bufsize > 0 ) {
 		// try and allocate the buffer
+		// note: the below always uses local allocation (and rightly so)
 		const int prc = posix_memalign(
 			&buffer,
 			grb::config::CACHE_LINE_SIZE::value(),
 			_bufsize
 		);
-		// Note: always uses local allocation (and rightly so)
 		if( prc != 0 ) {
 			delete[] cur_payload;
 			delete[] cur_msgs;
-#ifdef _DEBUG
-			std::cerr << "\t out of memory while initializing (3)\n";
-#endif
+			std::cerr << "\tError during BSP1D::init: "
+				<< "out of memory while initializing (3)\n";
 			return OUTOFMEM;
 		}
 	} else {
 		// set buffer to empty
-		buffer = NULL;
+		buffer = nullptr;
 	}
 
-#ifdef _DEBUG
-	std::cout << "\t " << s << ": calling lpf_resize_message_queue, requesting a "
-		<< "capacity of " << _maxh << ". Context is " << context << std::endl;
-#endif
+	// now make allocation requests of LPF:
 
-	// resize LPF buffers accordingly
-	lpf_err_t lpfrc = lpf_resize_message_queue( context, _maxh );
-	if( lpfrc == LPF_SUCCESS ) {
-		lpfrc = lpf_probe( context, &lpf_info );
+	// first, the machine info
+	lpf_err_t lpfrc = lpf_probe( context, &lpf_info );
+	if( lpfrc == LPF_ERR_OUT_OF_MEMORY ) {
+		delete [] cur_payload;
+		delete [] cur_msgs;
+		if( buffer != nullptr ) { free( buffer ); }
```

Comment:
this pattern occurs repeatedly throughout the code, and it is the typical "resource cleanup" pattern. It ultimately boils down to the "resource management" pattern, which C++ best implements via [RAII](https://en.cppreference.com/w/cpp/language/raii). Here, using `std::unique_ptr` would thus lead to less code written (these `delete []` statements would be unnecessary) and to more safety, as future modifications would not have deal with resource management explicitly and the code would become exception-safe automatically.

Reply:
Thanks! Lemme look at refactoring this after the below snafus are handled gone